std/assembly/array.ts

```diff
@@ -396,7 +396,8 @@ export class Array<T> {
 
       let out = result;
       if (estLen > offset) {
-        out = result.substring(0, offset);
+        out = allocateUnsafeString(offset);
+        copyUnsafeString(out, 0, result, 0, offset);
         freeUnsafeString(result);
```

Comment:
Is the `var result = "";` at the beginning necessary, with a new scoped `let result = allocateUnsafeString(estLen);` within here? Asking because the outer one is a GCed string due to initializing it with `""`, while the inner isn't, which immediately rang a bell here because the outer one can't be free'd.

Reply:
got it, I'll fix that

Reply:
(maybe rename the inner one to `buffer` for clarity?)

Reply:
I just removed it at all. It unnecessary now

Reply:
Now I wonder if that actually works, because `allocateUnsafeString` can create a GCed string as well, and `freeUnsafeString` only actually free's [if GC is not present](https://github.com/AssemblyScript/assemblyscript/blob/master/std/assembly/internal/string.ts#L26). `result` here should probably be a regular buffer and not be allocated as a string at all.

Reply:
Hmm, what you mean? result could return as well as `out`. See this for example: https://github.com/AssemblyScript/assemblyscript/blob/fd66274ca1af96c5b668c08563532a081e887fe5/std/assembly/array.ts#L516

So if result allocated by `allocateUnsafeString` much to actual produced length we just return it, but it less do substring and return truncated version.

Reply:
But wasn't the initial intent also to reduce GC overhead, like it is imposed by using `.substring`? If so, using `allocateUnsafeString` still imposes GC overhead due to creating a GCed object, and the proper solution would be to use `memory.allocate` instead, and only create a string once at the end through `allocateUnsafeString`. Might be missing something ofc

Reply:
I can create gc-unmanagable pure buffer in first step and always create new one via substring and return it. But sometimes this unnecessary. For example:
we have array like this:
```ts
var arr: bool[] = [false, false, false];
```
So we expect worse case (estimated length) and allocate pure buffer `strlen("false") * 3`, filled up it and create new one with same size but managed with gc and return it. So we do unnecessary job twice in this pretty often case. In current scenario we checking estimated length and actual length after filling first buffer and could skip next substring branch

Reply:
Just note that `freeUnsafeString` does nothing if a GC is present, hence there's a bit of GC overhead for releasing `result` at the next collection cycle. Not sure if this can be avoided?

Reply:
Actually I manage this. See [this](https://github.com/AssemblyScript/assemblyscript/blob/423ff6034db9e63844c49450c12db6e888af79e6/std/assembly/internal/string.ts#L27)

Reply:
This create overhead when we don't need truncate buffer as I mentioned before. So need investigate which is case better or worst. Hard to tell now.